### PR TITLE
Adds more syscall memory access checks

### PR DESCRIPTION
**Pull request by Sonicadvance1**
Does a decent number of syscalls, but not all of them. There's only so many I can do at a time before I start becoming mindnumb.

### CODE REVIEW

**Comment by neobrain** [comment]
Happy to wave this through since it's easy to refactor later, but here's an idea: Instead of having a two-step "verify"+"execute" that's potentially expensive, could we instead just toggle a thread-local flag that tells SignalDelegator to assume any SIGSEGV is triggered by invalid syscall arguments?

This would turn most code into something like this:
```
      if (Result != -1) {
        ExpectNoWriteFaults guard; // RAII wrapper around thread-local boolean

        // Any SIGSEGV will now be handled as if FaultSafeUserMemAccess::VerifyIsWritable failed.
        // It's important to keep the scope of the wrapper minimal hence

        for (size_t i = 0; i < Result; ++i) {
          events[i] = Events[i];
        }
        // ExpectNoWriteFaults destructor runs here and will unset the thread-local boolean
      }
```

**Comment by Sonicadvance1** [comment]
> Happy to wave this through since it's easy to refactor later, but here's an idea: Instead of having a two-step "verify"+"execute" that's potentially expensive, could we instead just toggle a thread-local flag that tells SignalDelegator to assume any SIGSEGV is triggered by invalid syscall arguments?
> 
> This would turn most code into something like this:
> 
> ```c++
>       if (Result != -1) {
>         ExpectNoWriteFaults guard; // RAII wrapper around thread-local boolean
> 
>         // Any SIGSEGV will now be handled as if FaultSafeUserMemAccess::VerifyIsWritable failed.
>         // It's important to keep the scope of the wrapper minimal hence
> 
>         for (size_t i = 0; i < Result; ++i) {
>           events[i] = Events[i];
>         }
>         // ExpectNoWriteFaults destructor runs here and will unset the thread-local boolean
>       }
> ```

We will also need to return EFAULT in the case that the fault does occur in the future. Not sure how that would operate but something to think about.

File: Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Msg.cpp [hunk 1]
```diff
@@ -80,17 +84,20 @@ void RegisterMsg(FEX::HLE::SyscallHandler* Handler) {
                               mq_attr* HostOld_p {};
 
                               if (newattr) {
+                                FaultSafeUserMemAccess::VerifyIsReadable(newattr, sizeof(*newattr));
                                 HostNew = *newattr;
                                 HostNew_p = &HostNew;
                               }
 
                               if (oldattr) {
+                                FaultSafeUserMemAccess::VerifyIsReadable(oldattr, sizeof(*oldattr));
```
Comment (neobrain) [inline Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Msg.cpp:93]: Does it need to be readable, though? We already check for writability below, and the variable isn't used before that.

File: Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Signals.cpp [hunk 2]
```diff
@@ -239,6 +250,8 @@ void RegisterSignals(FEX::HLE::SyscallHandler* Handler) {
       siginfo_t* info64_p {};
 
       if (info) {
+        FaultSafeUserMemAccess::VerifyIsReadable(info, sizeof(*info));
+        info64 = *info;
```
Comment (neobrain) [inline Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Signals.cpp:254]: What's this added line doing here?

File: Source/Tools/LinuxEmulation/LinuxSyscalls/x64/Semaphore.cpp [hunk 3]
```diff
@@ -29,8 +29,10 @@ void RegisterSemaphore(FEX::HLE::SyscallHandler* Handler) {
     case IPC_SET: {
       struct semid64_ds buf {};
       buf = *semun.buf;
+      FaultSafeUserMemAccess::VerifyIsReadable(semun.buf, sizeof(*semun.buf));
```
Comment (neobrain) [inline Source/Tools/LinuxEmulation/LinuxSyscalls/x64/Semaphore.cpp:32]: Check should be moved before the previous line.

**Comment by neobrain** [comment]
> We will also need to return EFAULT in the case that the fault does occur in the future. Not sure how that would operate but something to think about.

Ah yes, good point.

File: Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Signals.cpp (hunk 2 above)
Comment (Sonicadvance1) [reply Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Signals.cpp:254]: This one was actually a bug, we were forgetting to copy over the data when provided

File: Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Msg.cpp (hunk 1 above)
Comment (Sonicadvance1) [reply Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Msg.cpp]: oop, removed

File: Source/Tools/LinuxEmulation/LinuxSyscalls/x64/Semaphore.cpp (hunk 3 above)
Comment (Sonicadvance1) [reply Source/Tools/LinuxEmulation/LinuxSyscalls/x64/Semaphore.cpp:31]: Moved before

File: Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Signals.cpp (hunk 2 above)
Comment (neobrain) [reply Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Signals.cpp:254]: Could you move functional changes to a different PR (or at least patch) for visibility? Hiding these in a NFC-sounding change is a bit surprising.

Comment (Sonicadvance1) [reply Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Signals.cpp:254]: Opened #4015 for the fix

**Review by neobrain — APPROVED** [review]

File: Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Signals.cpp (hunk 2 above)
Comment (Sonicadvance1) [reply Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Signals.cpp:254]: Rebased

**Pull request merged**